### DISCUSSION TITLE
Speed Estimation - Ultralytics YOLOv8 Docs

### DISCUSSION THREAD

**Discussion by giscus** [ultralytics/ultralytics#7395] [Docs]
# Speed Estimation - Ultralytics YOLOv8 Docs

Speed Estimation Using Ultralytics YOLOv8

https://docs.ultralytics.com/guides/speed-estimation/

**Comment by AnirAnz456346** [comment]
Hi, on using the above code I'm getting this error.

ImportError: cannot import name 'speed_estimation' from 'ultralytics' (/home/anzo/.local/lib/python3.10/site-packages/ultralytics/__init__.py)

I've ensured that Ultralytics is already installed with pip, yet I'm still getting the error.
Any help would be much appreciated.

**Reply by RizwanMunawar** [reply]
@AnirAnz456346, appreciate you sharing the error logs. After conducting tests, it appears that the code is functioning correctly. Ensure you have the latest version of Ultralytics PYPI installed by using the provided command below.
```
pip install -U ultralytics
```

Hope this helps! Let me know if you have any additional questions.

Thanks and Regards
Ultralytics Team!

**Reply by AnirAnz456346** [reply]
Apparently, I had an older version installed. Re-installation solved the issue.
Thanks.

**Reply by pderrenger** [reply]
@AnirAnz456346 you're welcome! I'm glad to hear that updating to the latest version resolved the issue. If you have any more questions or need further assistance as you work with Ultralytics YOLOv8, don't hesitate to reach out. Happy coding and best of luck with your speed estimation project! 😊🚀

**Comment by yeongnamtan** [comment]
i followed the code above.  What I get is just the class of object and no speed information.  Am I missing something ? I used latest ultralytics version

**Reply by RizwanMunawar** [reply]
@yeongnamtan, the car's speed will be visible only after it crosses the plotted center line. We are actively enhancing the speed estimation module, and you can expect numerous updates in the upcoming days.

Thanks,
Ultralytics Team!

**Comment by Aenast28** [comment]
Hi, 
As my passion project, I'm trying to implement a solution where the user can click between two points and it creates a line, that will use the speed estimation calculation. This makes it really adjustable in real time.
While I've managed to implement it successfuly, I have two questions. 
1. For some reason, whenever the line is drawn diagonally or really in any acute angle, the estimations of speed are really inconsistent. Would anyone know why?
2. Does someone have any idea how to implement it such the code works with multiple reg_lines? Like if instead of one line, there would be multiple lines calculating the speed?

Thanks in advance!

**Reply by RizwanMunawar** [reply]
@Aenast28, great job! You can independently incorporate support for multiple region lines by modifying the module code. Find the code at the provided link: https://github.com/ultralytics/ultralytics/blob/main/ultralytics/solutions/speed_estimation.py

Thanks
Ultralytics Team!

**Reply by Aenast28** [reply]
And do you have any ideas or suggestion on how to do it?

**Reply by pderrenger** [reply]
@Aenast28 certainly! Here are a few suggestions to address your questions:

1. **Inconsistent Speed Estimations at Acute Angles:**
   - **Perspective Distortion:** At acute angles, the perspective distortion can affect the apparent distance between points in the image. You might need to apply perspective correction based on camera calibration to get accurate distance measurements.
   - **Scale Factor:** Consider implementing a scale factor that adjusts the speed estimation based on the angle of the line with respect to the camera's field of view.
   - **Frame Rate Consistency:** Ensure that the frame rate of the video is consistent, as fluctuations can affect speed calculations.

2. **Implementing Multiple `reg_lines`:**
   - **Data Structure:** Use a list of tuples or a dictionary to store multiple `reg_lines`, where each line has a unique identifier.
   - **Looping:** Modify the `speed_estimation` function to loop over each `reg_line` and perform speed calculations independently.
   - **Tracking:** Ensure that the object tracking is robust enough to track objects across different lines and maintain their identities.

Here's a rough outline of how you might modify the code to handle multiple `reg_lines`:

```python
class SpeedEstimator:
    def __init__(self):
        self.reg_lines = []  # List to store multiple region lines

    def add_reg_line(self, line_pts):
        self.reg_lines.append(line_pts)  # Add a new region line

    def estimate_speed(self, frame, tracks):
        for line_pts in self.reg_lines:
            # Perform speed estimation for each line
            # ...
        return frame

# Usage
speed_obj = SpeedEstimator()
speed_obj.add_reg_line([(0, 360), (1280, 360)])
speed_obj.add_reg_line([(0, 400), (1280, 400)])  # Add another line
# ...
```

Remember to handle the logic for each line separately, especially if the lines intersect or are close to each other. You may need to track which object crosses which line to avoid double-counting.

I hope these suggestions help you refine your speed estimation project! Keep up the great work! 🚀👨‍💻

**Comment by AhmedIbrahimai** [comment]
is there a code for speed estimation and count in one block ?

**Reply by RizwanMunawar** [reply]
@AhmedIbrahimai Both are different concepts. But you can modify the code according to your use case in the [Ultralytics Solutions Directory](https://github.com/ultralytics/ultralytics/tree/main/ultralytics/solutions)!

Thanks
Ultralytics Team!

**Comment by faliwew** [comment]
Hi, I'm placing a line horizontally, as the flow is the opposite of the example, do I have to change something in the code? Because I comment putting the line is not calculating the speed.

**Reply by pderrenger** [reply]
Hey there! 👋 It sounds like you're working on speed estimation and encountering some issues with the line placement. The direction of the flow shouldn't require any changes in the code aside from potentially adjusting the `line_pts` to match your scenario. The `line_pts` variable defines where the line is drawn for speed estimation. Make sure the coordinates for `line_pts` match the orientation and position of the line in your video. Here's a quick example:

```python
line_pts = [(start_x, start_y), (end_x, end_y)]  # Adjust these coordinates to your line placement
```

If you've adjusted `line_pts` and still face issues, double-check if the tracked objects are correctly passing over this line and consider adjusting the `conf` and `iou` thresholds for more accurate detection/tracking. Keep us posted on how it goes! 🚀

**Reply by faliwew** [reply]
Don't you need to change it here too? inside SpeedEstimator?
reg_pts (list): Initial list of points that define the velocity calculation region.

**Reply by pderrenger** [reply]
Hi there! 👋 Yes, if you're implementing or adjusting the speed estimation and wish to change the region where velocity is calculated, you indeed need to update the `reg_pts` in the `SpeedEstimator` initialization to reflect your new calculation region. Here's a quick code snippet on how to do it:

```python
from ultralytics.solutions import speed_estimation

# Define your new region of interest for speed calculation
new_reg_pts = [(new_start_x, new_start_y), (new_end_x, new_end_y)]

# Initialize SpeedEstimator with your custom reg_pts
speed_obj = speed_estimation.SpeedEstimator()
speed_obj.set_args(reg_pts=new_reg_pts, ...)
```

Just replace `new_reg_pts` with the coordinates of your new line. Happy coding and good luck with your speed estimation! 🚀😊
